cli/command/telemetry_utils.go

```diff
@@ -0,0 +1,172 @@
+//
+// Note: this should be the last func to wrap/modify the PersistentRunE/RunE funcs before command execution.
+//
+// can also be used for spans!
+func InstrumentCobraCommands(ctx context.Context, cmd *cobra.Command, meter metric.Meter) {
+	// If PersistentPreRunE is nil, make it execute PersistentPreRun and return nil by default
+	ogPersistentPreRunE := cmd.PersistentPreRunE
+	if ogPersistentPreRunE == nil {
+		ogPersistentPreRun := cmd.PersistentPreRun
+		ogPersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+			ogPersistentPreRun(cmd, args)
+			return nil
+		}
+		cmd.PersistentPreRun = nil
+	}
+
+	// wrap RunE in PersistentPreRunE so that this operation gets executed on all children commands
+	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		// If RunE is nil, make it execute Run and return nil by default
+		ogRunE := cmd.RunE
+		if ogRunE == nil {
+			ogRun := cmd.Run
+			ogRunE = func(cmd *cobra.Command, args []string) error {
+				ogRun(cmd, args)
+				return nil
+			}
+			cmd.Run = nil
+		}
+		cmd.RunE = func(cmd *cobra.Command, args []string) error {
+			// insert the meter and tracer objects into the context
+			// so commands can use them in their RunE func
+			newCtx := context.WithValue(cmd.Context(), OtelMeterCtxKey, meter)
+			// newCtx := context.WithValue(newCtx, OtelTracerCtxKey, tracer))
+			cmd.SetContext(newCtx)
+
+			// start the timer as the first step of every cobra command
+			stopCobraCmdTimer := startCobraCommandTimer(cmd)
+			cmdErr := ogRunE(cmd, args)
+			stopCobraCmdTimer(cmdErr)
+			return cmdErr
+		}
+
+		return ogPersistentPreRunE(cmd, args)
+	}
+}
+
+func startCobraCommandTimer(cmd *cobra.Command) func(err error) {
+	ctx := cmd.Context()
+	meter, ok := ctx.Value(OtelMeterCtxKey).(metric.Meter)
+	if meter == nil || !ok {
+		otel.Handle(errors.New("meter not found in ctx."))
+		// return a no-op if the meter cannot be found in the command's ctx
+		return func(err error) {}
+	}
+	baseAttrs := BaseMetricAttributes(cmd)
+	durationCounter, _ := meter.Float64Counter(
+		"command.duration",
+		metric.WithDescription("Measures the duration of the cobra command"),
+		metric.WithUnit("ms"),
+	)
+	start := time.Now()
+
+	return func(err error) {
+		duration := float64(time.Since(start)) / float64(time.Millisecond)
+		extraAttrs := []attribute.KeyValue{}
+		exitCode := 0
+		if err != nil {
+			exitCode = 1
+			if stderr, ok := err.(statusError); ok {
+				// StatusError should only be used for errors, and all errors should
+				// have a non-zero exit status, so only set this here if this value isn't 0
+				if stderr.StatusCode != 0 {
+					exitCode = stderr.StatusCode
+				}
+			}
+			// TODO(krissetto): this attribute might not be needed, but buildx has it. Check if it's necessary
+			extraAttrs = append(extraAttrs, attribute.String("command.error.type", otelErrorType(err)))
```

Comment:
Can you consolidate this into a function so it's just something like `exitCodeFromError`?

Also none of these attributes are really _needed_, but I liked error type because it makes it potentially easier to categorize types of errors later. In buildx, we're just using it right now to determine if the command was successful, canceled, or anything else.

Reply:
I consolidated this logic into an `attributesFromError` func which return a series of err-dependent attributes. 
IMHO any extra refactoring of this part should be done once we start implementing more logic and our needs are clearer